File: src/utils/bilibiliLiveUtils.ts
import { invoke } from "@tauri-apps/api/core";
import { parseCookies } from "./defaultUtils";

interface wsPacket {
    Len?: number;
    HeadLen?: number;
    Ver?: number;
    Type?: number;
    Num?: number;
    body?: string | number;
}

function randomElement<T>(list: T[]): T {
    return list[Math.floor(Math.random() * list.length)]
}

export const webSocket = async (room_id: number, cookie?: string): Promise<void> => {
    if ("WebSocket" in window) {
        console.log("您的浏览器支持WebSocket");
        var timer: number | null = null;
        const room = await invoke<any>('proxy_request', { url: `https://api.live.bilibili.com/room/v1/Room/get_info?room_id=${room_id}` })
            .then(res => res)
            .catch(err => {
                console.log(`获取房间ID失败: ${err as string}`);
                return;
            });
        const damuInfoUrl = `https://api.live.bilibili.com/xlive/web-room/v1/index/getDanmuInfo?id=${room.data.room_id}`;
        let danmuInfo: any;
        let uid: number;
        let buvid: string;

        if (cookie) {
            const cookies = parseCookies(cookie);
            uid = parseInt(cookies.DedeUserID);
            buvid = cookies.buvid3;
            await invoke<any>('proxy_request', { url: damuInfoUrl, cookie: cookie })
                .then(response => {
                    danmuInfo = response;
                })
                .catch(err => {
                    console.log(`获取弹幕服务器配置失败: ${err as string}`);
                    return;
                });
        } else {
            await invoke<any>('proxy_request', { url: damuInfoUrl })
                .then(response => {
                    danmuInfo = response;
                })
                .catch(err => {
                    console.log(`获取弹幕服务器配置失败: ${err as string}`);
                    return;
                });
            uid = 0;
            buvid = "";
        }

        const host: any = randomElement(danmuInfo.data.host_list)
        const url = `wss://${host.host}:${host.wss_port}/sub`

        var certification = {
            "uid": uid,
            "roomid": room.data.room_id,
            "protover": 3,
            "buvid": buvid,
            "platform": "web",
            "type": 2,
            "key": danmuInfo.data.token  //值为空字符串好像也没问题
        }
        console.log(JSON.stringify(certification))

        var ws = new WebSocket(url);

        ws.onopen = function (e) {
            console.log("open");
            ws.send(getCertification(JSON.stringify(certification)));
            //发送心跳包
            timer = setInterval(function () {
                let buff = new ArrayBuffer(16);
                let i = new DataView(buff);
                i.setUint32(0, 16);    //整个封包
                i.setUint16(4, 16);    //头部
                i.setUint16(6, 1);    //协议版本
                i.setUint32(8, 2);    //操作码,2为心跳包
                i.setUint32(12, 1);    //填1
                ws.send(buff);
            }, 30000); //30秒

        }

        ws.onmessage = function (e) {
            let blob = e.data;
            let decoder = new TextDecoder();
            let reader = new FileReader();
            reader.readAsArrayBuffer(blob);
            reader.onload = async function () {
                let buff = reader.result as ArrayBuffer;
                let view = new DataView(buff);
                let offset = 0;
                let packet: wsPacket = {};
                // let result: wsPacket[] = [];
                while (offset < buff.byteLength) {
                    let packetLen = view.getUint32(offset + 0);
                    let headLen = view.getUint16(offset + 4);
                    let packetVer = view.getUint16(offset + 6);
                    let packetType = view.getUint32(offset + 8);
                    let num = view.getUint32(12);
                    let dataArray = new Uint8Array(buff, offset + headLen, packetLen - headLen);
                    if (packetVer === 3) {
                        await invoke<Uint8Array>('decompress_brotli', { data: dataArray }).then(data => {
                            let uint8Array = new Uint8Array(data);
                            let arrayBuffer = uint8Array.buffer;
                            let view = new DataView(arrayBuffer);
                            let offset_Ver3 = 0;
                            while (offset_Ver3 < arrayBuffer.byteLength) {
                                let packetLen = view.getUint32(offset_Ver3 + 0);
                                let headLen = view.getUint16(offset_Ver3 + 4);
                                let packetVer = view.getUint16(offset_Ver3 + 6);
                                let packetType = view.getUint32(offset_Ver3 + 8);
                                let num = view.getUint32(12);
                                packet.Len = packetLen;
                                packet.HeadLen = headLen;
                                packet.Ver = packetVer;
                                packet.Type = packetType;
                                packet.Num = num;
                                let dataArray = new Uint8Array(arrayBuffer, offset_Ver3 + headLen, packetLen - headLen);
                                packet.body = JSON.parse(decoder.decode(dataArray));
                                offset_Ver3 += packetLen;
                            }
                        }).catch(err => {
                            console.error('解压数据失败:', err);
                            return new Uint8Array();
                        })
                    } else {
                        packet.Len = packetLen;
                        packet.HeadLen = headLen;
                        packet.Ver = packetVer;
                        packet.Type = packetType;
                        packet.Num = num;
                        if (packetType == 3) {
                            packet.body = (new DataView(buff, offset + headLen, packetLen - headLen)).getUint32(0);
                        } else {
                            packet.body = JSON.parse(decoder.decode(dataArray));
                        }
                    }
                    console.log(packet.body)
                    offset += packetLen;
                }


            }
            // handleMessage(blob, function (result: string[]) {
            //     //触发事件
            //     for (let i = 0; i < result.length; i++) {
            //         let json = JSON.parse(result[i]);
            //         if (json.Type == 5) {
            //             let event = new CustomEvent(JSON.parse(json.body).cmd, { detail: JSON.parse(json.body) });
            //             eventTarget.dispatchEvent(event);
            //         }
            //         if (json.Type == 8) {
            //             let event = new CustomEvent("Certify_Success", { detail: JSON.parse(json.body) });
            //             eventTarget.dispatchEvent(event);
            //         }
            //         if (json.Type == 3) {
            //             let event = new CustomEvent("VIEW", { detail: json.body });
            //             eventTarget.dispatchEvent(event);
            //         }
            //     }
            // });
        }

        ws.onclose = function (e) {
            //当客户端收到服务端发送的关闭连接请求时，触发onclose事件
            console.log("close");
            if (timer != null) {
                clearInterval(timer);    //停止发送心跳包
            }
            // setTimeout(() => webSocket(room_id), 4000);    //4秒后重连
        }

        ws.onerror = function (e) {
            //如果出现连接、处理、接收、发送数据失败的时候触发onerror事件
            console.log(e);
        }
    } else {
        console.log("您的浏览器不支持WebSocket");
    }
}

//生成认证数据
const getCertification = (json: string): ArrayBuffer => {
    let encoder = new TextEncoder();    //编码器
    let jsonView = encoder.encode(json);    //utf-8编码
    let buff = new ArrayBuffer(jsonView.byteLength + 16);    //数据包总长度：16位头部长度+bytes长度
    let view = new DataView(buff);    //新建操作视窗
    view.setUint32(0, jsonView.byteLength + 16);    //整个数据包长度
    view.setUint16(4, 16);    //头部长度
    view.setUint16(6, 1);    //协议版本
    view.setUint32(8, 7);    //类型,7为加入房间认证
    view.setUint32(12, 1);    //填1
    for (let r = 0; r < jsonView.byteLength; r++) {
        view.setUint8(16 + r, jsonView[r]);    //填入数据
    }
    return buff;
}

File: src-tauri/src/utils/bilibili_live_utils.rs
use brotli::BrotliDecompress;
use futures::{SinkExt, StreamExt};
use rand::seq::SliceRandom;
use serde_json::{json, Value};
use std::collections::HashMap;
use std::io::Cursor;
use std::time::Duration;

pub struct WsPacket {
    pub len: Option<i32>,
    pub head_len: Option<i32>,
    pub ver: Option<i32>,
    pub packet_type: Option<i32>,
    pub num: Option<i32>,
    pub body: Option<String>,
}

pub struct bilibiliWs {
    live_id: String,
    cookie: Option<String>,
    callbacks: HashMap<String, Box<dyn Fn(&Value) + Send + Sync>>,
}

impl bilibiliWs {
    pub fn new(live_id: &str, cookie: Option<String>) -> Self {
        bilibiliWs {
            live_id: live_id.to_string(),
            cookie,
            callbacks: HashMap::new(),
        }
    }

    pub async fn start(&mut self) {
        let client = reqwest::Client::new();
        let url = format!(
            "https://api.live.bilibili.com/room/v1/Room/get_info?room_id={}",
            self.live_id
        );
        let response = client.get(&url).send().await.expect("获取房间信息失败");
        let room_info: Value = response.json().await.expect("解析JSON失败");
        let room_id = room_info["data"]["room_id"]
            .as_i64()
            .expect("获取房间ID失败");
        let danmu_info_url = format!(
            "https://api.live.bilibili.com/xlive/web-room/v1/index/getDanmuInfo?id={}",
            room_id
        );
        let mut uid = 0;
        let mut buvid = "";
        let danmu_info = match &self.cookie {
            Some(cookie) => {
                let response = client
                    .get(&danmu_info_url)
                    .header("Cookie", cookie)
                    .send()
                    .await
                    .expect("获取弹幕服务器配置失败")
                    .json::<Value>()
                    .await
                    .expect("解析弹幕服务器配置失败");

                let cookies = cookie.split(';').collect::<Vec<_>>();
                uid = cookies
                    .iter()
                    .find(|c| c.trim().starts_with("DedeUserID="))
                    .and_then(|c| c.split('=').nth(1))
                    .and_then(|id| id.parse::<i64>().ok())
                    .unwrap_or(0);
                buvid = cookies
                    .iter()
                    .find(|c| c.trim().starts_with("buvid3="))
                    .and_then(|c| c.split('=').nth(1))
                    .unwrap_or("");
                response
            }
            None => client
                .get(&danmu_info_url)
                .send()
                .await
                .expect("获取弹幕服务器配置失败")
                .json::<Value>()
                .await
                .expect("解析弹幕服务器配置失败"),
        };
        let host = danmu_info["data"]["host_list"]
            .as_array()
            .expect("获取弹幕服务器列表失败")
            .choose(&mut rand::thread_rng())
            .expect("选择弹幕服务器失败");
        let url = format!(
            "wss://{}:{}/sub",
            host["host"].as_str().expect("获取host失败"),
            host["wss_port"].as_i64().expect("获取端口失败")
        );

        let certification = json!({
            "uid": uid,
            "roomid": room_id,
            "protover": 3,
            "buvid": buvid,
            "platform": "web",
            "type": 2,
            "key": danmu_info["data"]["token"].as_str().expect("获取token失败")
        });

        let (ws_stream, _) = tokio_tungstenite::connect_async(url)
            .await
            .expect("连接失败");
        let (mut write, mut read) = ws_stream.split();

        // 发送认证包
        let mut buff = Vec::with_capacity(certification.to_string().len() + 16);
        // 整个数据包长度
        buff.extend_from_slice(&((certification.to_string().len() + 16) as u32).to_be_bytes());
        // 头部长度
        buff.extend_from_slice(&16u16.to_be_bytes());
        // 协议版本
        buff.extend_from_slice(&1u16.to_be_bytes());
        // 类型,7为加入房间认证
        buff.extend_from_slice(&7u32.to_be_bytes());
        // 填1
        buff.extend_from_slice(&1u32.to_be_bytes());
        // 填入数据
        buff.extend_from_slice(certification.to_string().as_bytes());
        write
            .send(tokio_tungstenite::tungstenite::Message::Binary(buff))
            .await
            .expect("发送认证包失败");

        tokio::spawn(async move {
            let mut interval = tokio::time::interval(Duration::from_secs(30));
            loop {
                interval.tick().await;
                let mut buff = Vec::with_capacity(16);
                // 整个封包长度
                buff.extend_from_slice(&16u32.to_be_bytes());
                // 头部长度
                buff.extend_from_slice(&16u16.to_be_bytes());
                // 协议版本
                buff.extend_from_slice(&1u16.to_be_bytes());
                // 操作码,2为心跳包
                buff.extend_from_slice(&2u32.to_be_bytes());
                // 填1
                buff.extend_from_slice(&1u32.to_be_bytes());
                write
                    .send(tokio_tungstenite::tungstenite::Message::Binary(buff))
                    .await
                    .expect("发送心跳包失败");
            }
        });

        // 处理接收到的消息
        while let Some(msg) = read.next().await {
            match msg {
                Ok(msg) => {
                    let data = msg.into_data();
                    let mut offset = 0;
                    while offset < data.len() {
                        let packet_len =
                            u32::from_be_bytes(data[offset..offset + 4].try_into().unwrap())
                                as usize;
                        let head_len =
                            u16::from_be_bytes(data[offset + 4..offset + 6].try_into().unwrap())
                                as usize;
                        let packet_ver =
                            u16::from_be_bytes(data[offset + 6..offset + 8].try_into().unwrap());
                        let _packet_type =
                            u32::from_be_bytes(data[offset + 8..offset + 12].try_into().unwrap());
                        let _num =
                            u32::from_be_bytes(data[offset + 12..offset + 16].try_into().unwrap());
                        let data_array = &data[offset + head_len..offset + packet_len];
                        if packet_ver == 3 {
                            let mut decompressed = Vec::new();
                            let mut reader = Cursor::new(data_array);
                            BrotliDecompress(&mut reader, &mut decompressed)
                                .expect("brotli解压失败");
                            let mut offset_ver3 = 0;
                            while offset_ver3 < decompressed.len() {
                                let packet_len = u32::from_be_bytes(
                                    decompressed[offset_ver3..offset_ver3 + 4]
                                        .try_into()
                                        .unwrap(),
                                ) as usize;
                                let head_len = u16::from_be_bytes(
                                    decompressed[offset_ver3 + 4..offset_ver3 + 6]
                                        .try_into()
                                        .unwrap(),
                                ) as usize;
                                let _packet_ver = u16::from_be_bytes(
                                    decompressed[offset_ver3 + 6..offset_ver3 + 8]
                                        .try_into()
                                        .unwrap(),
                                );
                                let _packet_type = u32::from_be_bytes(
                                    decompressed[offset_ver3 + 8..offset_ver3 + 12]
                                        .try_into()
                                        .unwrap(),
                                );
                                let _num = u32::from_be_bytes(
                                    decompressed[offset_ver3 + 12..offset_ver3 + 16]
                                        .try_into()
                                        .unwrap(),
                                );
                                let data_array =
                                    &decompressed[offset_ver3 + head_len..offset_ver3 + packet_len];
                                let body = String::from_utf8_lossy(data_array);
                                let json_value: Value = serde_json::from_str(&body).unwrap();
                                let cmd = json_value["cmd"].as_str().unwrap();
                                if let Some(callback) = self.callbacks.get(cmd) {
                                    match cmd {
                                        "DANMU_MSG" => {
                                            let json_msg = json!({
                                                "user": json_value["info"][2][1].as_str().unwrap(),
                                                "content": json_value["info"][1].as_str().unwrap(),
                                            });
                                            callback(&json_msg);
                                        }
                                        "SEND_GIFT" => {
                                            let json_msg = json!({
                                                "user": json_value["data"]["uname"].as_str().unwrap(),
                                                "gift_name": json_value["data"]["giftName"].as_str().unwrap(),
                                                "count": json_value["data"]["num"].as_i64().unwrap(),
                                            });
                                            callback(&json_msg);
                                        }
                                        "INTERACT_WORD" => {
                                            let json_msg = json!({
                                                "user": json_value["data"]["uname"].as_str().unwrap(),
                                                "msg_type": json_value["data"]["msg_type"].as_i64().unwrap(),
                                            });
                                            callback(&json_msg);
                                        }
                                        "LIKE_INFO_V3_CLICK" => {
                                            let json_msg = json!({
                                                "user": json_value["data"]["uname"].as_str().unwrap(),
                                            });
                                            callback(&json_msg);
                                        }
                                        _ => {}
                                    }
                                }
                                offset_ver3 += packet_len;
                            }
                        } else {
                            // if packet_type == 3 {
                            //     let body = u32::from_be_bytes(data_array[0..4].try_into().unwrap());
                            // } else {
                            //     let body = String::from_utf8_lossy(data_array);
                            // }
                        }
                        offset += packet_len;
                    }
                }
                Err(e) => {
                    println!("接收消息错误: {:?}", e);
                }
            }
        }
    }

    pub fn on<F>(&mut self, message_type: &str, callback: F)
    where
        F: Fn(&Value) + Send + Sync + 'static,
    {
        self.callbacks
            .insert(message_type.to_string(), Box::new(callback));
    }
}
